File: C#/IDX+IMG.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using GovanifY.Utility;
using KH2FM_Toolkit;
using KHCompress;

namespace IDX_Tools
{
    internal class IDXFile : IDisposable, IEnumerable<IDXFile.IDXEntry>
    {
        private readonly bool leaveOpen;
        private BinaryReader file;

        protected IDXFile()
        {
            file = null;
        }

        public IDXFile(Stream input, bool newidx = false, bool leaveOpen = false)
        {
            file = new BinaryReader(input);
            this.leaveOpen = leaveOpen;
            input.Position = 0;
            if (newidx)
            {
                Count = 0;
                input.Write(new byte[] {0, 0, 0, 0}, 0, 4);
            }
            else
            {
                Count = file.ReadUInt32();
            }
            Position = 0;
        }

        public uint Count { get; private set; }
        public uint Position { get; private set; }

        public void Dispose()
        {
            if (file != null)
            {
                if (!leaveOpen)
                {
                    file.Close();
                }
                file = null;
            }
        }

        public IEnumerator<IDXEntry> GetEnumerator()
        {
            for (uint i = 0; i < Count; ++i)
            {
                yield return ReadEntry(i);
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public IDXEntry ReadEntry(long index = -1)
        {
            if (index >= 0)
            {
                Position = (uint) index;
            }
            if (Position >= Count)
            {
                return new IDXEntry {Hash = 0};
            }
            file.BaseStream.Position = 4 + 16*Position++;
            var entry = new IDXEntry
            {
                Hash = file.ReadUInt32(),
                HashAlt = file.ReadUInt16(),
                Compression = file.ReadUInt16(),
                DataLBA = file.ReadUInt32(),
                DataLength = file.ReadUInt32()
            };
            return entry;
        }

        private void WriteEntry(IDXEntry entry, uint count = 0)
        {
            using (var bw = new BinaryStream(file.BaseStream, leaveOpen: true))
            {
                bw.Write(entry.Hash);
                bw.Write(entry.HashAlt);
                bw.Write(entry.Compression);
                bw.Write(entry.DataLBA);
                bw.Write(entry.DataLength);
                if (count != 0)
                {
                    bw.BaseStream.Position = 0;
                    bw.Write(count);
                }
            }
        }

        private void FindEntryByHash(uint hash)
        {
            file.BaseStream.Position = 4;
            for (uint i = 0; i < Count; ++i)
            {
                if (file.ReadUInt32() == hash)
                {
                    file.BaseStream.Position -= 4;
                    return;
                }
                file.BaseStream.Position += 12;
            }
            throw new FileNotFoundException();
        }

        /// <summary>Relink hash to target</summary>
        /// <param name="hash"></param>
        /// <param name="target"></param>
        public void RelinkEntry(uint hash, uint target)
        {
            FindEntryByHash(target);
            file.BaseStream.Position += 4;
            var entry = new IDXEntry
            {
                Hash = hash,
                HashAlt = file.ReadUInt16(),
                Compression = file.ReadUInt16(),
                DataLBA = file.ReadUInt32(),
                DataLength = file.ReadUInt32()
            };
            FindEntryByHash(hash);
            WriteEntry(entry);
        }

        public void ModifyEntry(IDXEntry entry)
        {
            FindEntryByHash(entry.Hash);
            WriteEntry(entry);
        }

        public void AddEntry(IDXEntry entry)
        {
            file.BaseStream.Position = 4 + 16*Count;
            WriteEntry(entry, ++Count);
        }

        public class IDXEntry
        {
            /// <summary>
            ///     <para>Compressed length of data, in sectors</para>
            ///     <para>Bit 0x4000 flags if compressed</para>
            /// </summary>
            public ushort Compression = 0;

            /// <summary>Location of data in IMG (LBA)</summary>
            public uint DataLBA;

            /// <summary>Data length (size of file)</summary>
            public uint DataLength;

            /// <summary>
            ///     <para>Identifier of this file</para>
            ///     <para>Hashed filename</para>
            /// </summary>
            public uint Hash;

            /// <summary>
            ///     <para>Secondary identifier of this file</para>
            ///     <para>Hashed filename</para>
            /// </summary>
            public ushort HashAlt = 0;

            /// <summary>Location of data in IMG (bytes)</summary>
            public long Offset
            {
                get { return DataLBA*2048; }
            }

            /// <summary>Returns true if this file is compressed</summary>
            public bool IsCompressed
            {
                get { return (Compression & 0x4000u) == 0x4000u; }
                set { Compression = (ushort) ((value ? 0x4000u : 0) | (Compression & 0xBFFFu)); }
            }

            /// <summary>Returns true if both hashes are checked</summary>
            public bool IsDualHash
            {
                get { return (Compression & 0x8000u) == 0x8000u; }
                set { Compression = (ushort) ((value ? 0x8000u : 0) | (Compression & 0x7FFFu)); }
            }

            /// <summary>Calculated the compressed data size, in bytes</summary>
            public uint CompressedDataLength
            {
                get
                {
                    // KH2 does get this value by and-ing with 0x3FFF, so that's confirmed
                    uint size = (Compression & 0x3FFFu) + 1u;
                    //Fixes file 0x10303F6F:
                    //  Real compressed size = 12,093,440 bytes (Verified manually)
                    //  Compression = 0x4710
                    //  (0x4710 & 0x3FFF) + 1 = 0x711 * 2048 = 3,704,832 + 0x800000 = 12,093,440 bytes
                    //What is happening is the size is getting truncated (see the set function).
                    //0x10303F6F is the only compressed file that hits this limitation officially, and uncompressed files can bypass it (use DataLength).
                    //Also note this increases *all* files about the specified size
                    if (Hash == 0x10303F6F && IsCompressed && DataLength > 0xC00000)
                    {
                        size += 0x1000u;
                    }
                    return size*2048u;
                }
                set
                {
                    if (value == 0)
                    {
                        Compression &= (ushort) 0xC000u;
                        return;
                    }
                    uint size = (uint) Math.Ceiling((double) value/2048) - 1;
                    //Seems that the "size" component just truncated if it's too large, as seen by the larger files (videos, Title.vas)
                    /* Size component:
                     * * Includes at minimum 0x0FFF
                     * * Does NOT include 0x1000 (@noname/10303F6F.bin; vagstream/Title.vas; zmovie/fm/me3.pss; zmovie/fm/opn.pss)
                     * * Does include 0x2000 (zmovie/fm/me3.pss; zmovie/fm/opn.pss)
                     * * 0x4000 is the compression flag
                     * * 0x8000 is the alt hash flag
                     * Nothing has broken these rules, that I can find
                    */
                    Compression = (ushort) ((Compression & 0xC000u) | (size & 0x2FFFu));
                }
            }
        }
    }

    internal class IDXFileWriter
    {
        private readonly List<IDXFile.IDXEntry> entries = new List<IDXFile.IDXEntry>();

        public void AddEntry(IDXFile.IDXEntry entry)
        {
            entries.Add(entry);
        }

        public void RelinkEntry(uint hash, uint target)
        {
            IDXFile.IDXEntry t = entries.Find(e => e.Hash == target);
            if (t.Hash == 0)
            {
                throw new FileNotFoundException();
            }
            entries.Add(new IDXFile.IDXEntry
            {
                Hash = hash,
                HashAlt = 0,
                Compression = t.Compression,
                DataLBA = t.DataLBA,
                DataLength = t.DataLength
            });
        }

        [SuppressMessage("Microsoft.Usage", "CA2202:Do not dispose objects multiple times")]
        public MemoryStream GetStream()
        {
            entries.Sort((a, b) => a.Hash < b.Hash ? -1 : (a.Hash > b.Hash ? 1 : 0));
            var ms = new MemoryStream();
            try
            {
                using (var bw = new BinaryStream(ms, leaveOpen: true))
                {
                    bw.Write(entries.Count);
                    foreach (IDXFile.IDXEntry entry in entries)
                    {
                        bw.Write(entry.Hash);
                        bw.Write(entry.HashAlt);
                        bw.Write(entry.Compression);
                        bw.Write(entry.DataLBA);
                        bw.Write(entry.DataLength);
                    }
                }
            }
            catch (Exception)
            {
                ms.Close();
                throw;
            }
            ms.Position = 0;
            return ms;
        }
    }

    internal class IMGFile : IDisposable
    {
        private static readonly byte[] copyBuffer = new byte[2048*2];
        private readonly bool leaveOpen;
        private readonly long offset;
        internal Stream file;

        public IMGFile(Stream file, long offset = 0, bool leaveOpen = false)
        {
            this.file = file;
            this.offset = offset;
            this.leaveOpen = leaveOpen;
        }

        public void Dispose()
        {
            Dispose(true);
        }

        public void Dispose(bool disposing)
        {
            if (file != null)
            {
                if (!leaveOpen)
                {
                    file.Dispose();
                }
                file = null;
            }
        }

        private void ReadFileBuffer(Stream destination, long origin, uint length)
        {
            //Not thread safe, but I'm not using threads
            file.Position = offset + origin;
            int num;
            while (length > 0 && (num = file.Read(copyBuffer, 0, (int) Math.Min(2048*2, length))) != 0)
            {
                destination.Write(copyBuffer, 0, num);
                length -= (uint) num;
            }
        }

        /// <summary>Ensure position is at a 2048 boundary</summary>
        private void EnsureBoundary()
        {
            if (((file.Position - offset)%2048) != 0)
            {
                int rem = 2048 - (int) ((file.Position - offset)%2048);
                byte[] buf = {0, 0, 0, 0};
                while (rem > 3)
                {
                    file.Write(buf, 0, 4);
                    rem -= 4;
                }
                while (--rem >= 0)
                {
                    file.WriteByte(0);
                }
            }
            if (((file.Position - offset)%2048) != 0)
            {
                throw new DataMisalignedException();
            }
        }

        public Substream GetFileStream(IDXFile.IDXEntry entry)
        {
            return new Substream(file, offset + entry.DataLBA*2048,
                entry.IsCompressed ? entry.CompressedDataLength : entry.DataLength);
        }

        public void Seek(uint sector)
        {
            file.Position = offset + sector*2048;
        }
        public void ReadFile(IDXFile.IDXEntry entry, Stream target, bool AdSize)
        {
            if (entry.IsCompressed)
            {
                if (entry.CompressedDataLength > int.MaxValue)
                {
                    throw new NotSupportedException("File to big to decompress");
                }
                var input = new byte[entry.CompressedDataLength];
                Seek(entry.DataLBA);
                file.Read(input, 0, (int) entry.CompressedDataLength);
                try
                {
                    byte[] output = KH2Compressor.decompress(input, entry.DataLength);
                    target.Write(output, 0, output.Length);
                    if (AdSize)
                    {
                        Console.WriteLine("Size (unpacked): {0}\r\n", output.Length);
                    }
                }
                catch (Exception e)
                {
                    Program.WriteError(" ERROR: Failed to decompress: " + e.Message);
                }
            }
            else
            {
                ReadFileBuffer(target, entry.Offset, entry.DataLength);
            }
        }

        public void WriteFile(Stream data)
        {
            if (data.Length > 0xFFFFFFFF)
            {
                throw new NotSupportedException("data too big to store");
            }
            EnsureBoundary();
            data.CopyTo(file);
            EnsureBoundary();
        }

        public void AppendFile(IDXFile.IDXEntry entry, Stream data)
        {
            if (data.Length > 0xFFFFFFFF)
            {
                throw new NotSupportedException("data too big to store");
            }
            file.Seek(0, SeekOrigin.End);
            EnsureBoundary();
            entry.DataLBA = (uint) (file.Position - offset)/2048;
            data.CopyTo(file);
            EnsureBoundary();
        }
    }

    internal class IDXIMGWriter : IDisposable
    {
        private IDXFileWriter idx = new IDXFileWriter();

        public IDXIMGWriter(Stream img, long imgoffset = 0, bool leaveOpen = false)
        {
            this.img = new IMGFile(img, imgoffset);
            this.img = new IMGFile(img, imgoffset, leaveOpen);
        }

        public IMGFile img { get; private set; }

        public void Dispose()
        {
            Dispose(true);
        }

        public void Dispose(bool disposing)
        {
            if (img != null)
            {
                img.Dispose();
                img = null;
            }
            if (disposing && idx != null)
            {
                idx = null;
            }
        }

        public void AddFile(IDXFile.IDXEntry file, Stream data)
        {
            img.AppendFile(file, data);
            idx.AddEntry(file);
        }

        public void RelinkFile(uint hash, uint target)
        {
            idx.RelinkEntry(hash, target);
        }

        public MemoryStream GetCurrentIDX()
        {
            return idx.GetStream();
        }
    }
}

File: C++/IDX+IMG.cpp
#include "IDX+IMG.h"

//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Collections;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Collections::Generic;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Diagnostics::CodeAnalysis;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::IO;
using namespace GovanifY::Utility;
using namespace KH2FM_Toolkit;
using namespace KHCompress;

namespace IDX_Tools
{

	const long long &IDXFile::IDXEntry::getOffset() const
	{
		return DataLBA*2048;
	}

	const bool &IDXFile::IDXEntry::getIsCompressed() const
	{
		return (Compression & 0x4000u) == 0x4000u;
	}

	void IDXFile::IDXEntry::setIsCompressed(const bool &value)
	{
		Compression = static_cast<unsigned short>((value ? 0x4000u : 0) | (Compression & 0xBFFFu));
	}

	const bool &IDXFile::IDXEntry::getIsDualHash() const
	{
		return (Compression & 0x8000u) == 0x8000u;
	}

	void IDXFile::IDXEntry::setIsDualHash(const bool &value)
	{
		Compression = static_cast<unsigned short>((value ? 0x8000u : 0) | (Compression & 0x7FFFu));
	}

	const unsigned int &IDXFile::IDXEntry::getCompressedDataLength() const
	{
		// KH2 does get this value by and-ing with 0x3FFF, so that's confirmed
		unsigned int size = (Compression & 0x3FFFu) + 1u;
		//Fixes file 0x10303F6F:
		//  Real compressed size = 12,093,440 bytes (Verified manually)
		//  Compression = 0x4710
		//  (0x4710 & 0x3FFF) + 1 = 0x711 * 2048 = 3,704,832 + 0x800000 = 12,093,440 bytes
		//What is happening is the size is getting truncated (see the set function).
		//0x10303F6F is the only compressed file that hits this limitation officially, and uncompressed files can bypass it (use DataLength).
		//Also note this increases *all* files about the specified size
		if (Hash == 0x10303F6F && getIsCompressed() && DataLength > 0xC00000)
		{
			size += 0x1000u;
		}
		return size*2048u;
	}

	void IDXFile::IDXEntry::setCompressedDataLength(const unsigned int &value)
	{
		if (value == 0)
		{
			Compression &= static_cast<unsigned short>(0xC000)u;
			return;
		}
		unsigned int size = static_cast<unsigned int>(ceil(static_cast<double>(value) / 2048)) - 1;
		//Seems that the "size" component just truncated if it's too large, as seen by the larger files (videos, Title.vas)
		/* Size component:
		 * * Includes at minimum 0x0FFF
		 * * Does NOT include 0x1000 (obj/B_LK120_RAW.mset; vagstream/Title.vas; zmovie/fm/me3.pss; zmovie/fm/opn.pss)
		 * * Does include 0x2000 (zmovie/fm/me3.pss; zmovie/fm/opn.pss)
		 * * 0x4000 is the compression flag
		 * * 0x8000 is the alt hash flag
		 * Nothing has broken these rules, that I can find
		*/
		Compression = static_cast<unsigned short>((Compression & 0xC000u) | (size & 0x2FFFu));
	}

	void IDXFile::IDXEntry::InitializeInstanceFields()
	{
		Compression = 0;
		HashAlt = 0;
	}

	IDXFile::IDXFile()
	{
//C# TO C++ CONVERTER WARNING: C# to C++ Converter converted the original 'null' assignment to a call to 'delete', but you should review memory allocation of all pointer variables in the converted code:
		delete file;
	}

	IDXFile::IDXFile(Stream *input, bool newidx = false, bool leaveOpen = false) : leaveOpen(leaveOpen)
	{
		file = new BinaryReader(input);
		input->Position = 0;
		if (newidx)
		{
			setCount(0);
			input->Write(new unsigned char[] {0, 0, 0, 0}, 0, 4);
		}
		else
		{
			setCount(file->ReadUInt32());
		}
		setPosition(0);
	}

	const unsigned int &IDXFile::getCount() const
	{
		return privateCount;
	}

	void IDXFile::setCount(const unsigned int &value)
	{
		privateCount = value;
	}

	const unsigned int &IDXFile::getPosition() const
	{
		return privatePosition;
	}

	void IDXFile::setPosition(const unsigned int &value)
	{
		privatePosition = value;
	}

	IDXFile::~IDXFile()
	{
		if (file != 0)
		{
			if (!leaveOpen)
			{
				file->Close();
			}
//C# TO C++ CONVERTER WARNING: C# to C++ Converter converted the original 'null' assignment to a call to 'delete', but you should review memory allocation of all pointer variables in the converted code:
			delete file;
		}
	}

	IEnumerator<IDXEntry*> *IDXFile::GetEnumerator()
	{
		for (unsigned int i = 0; i < getCount(); ++i)
		{
//C# TO C++ CONVERTER TODO TASK: C++ does not have an equivalent to the C# 'yield' keyword:
			yield return ReadEntry(i);
		}
	}

	IEnumerator *IDXFile::IEnumerable_GetEnumerator()
	{
		return GetEnumerator();
	}

	IDX_Tools::IDXFile::IDXEntry *IDXFile::ReadEntry(long long index = -1)
	{
		if (index >= 0)
		{
			setPosition(static_cast<unsigned int>(index));
		}
		if (getPosition() >= getCount())
		{
			return new IDXEntry {Hash = 0};
		}
		file->BaseStream->Position = 4 + 16*getPosition();
	setPosition(getPosition() + 1);
		IDXEntry *entry = new IDXEntry {Hash = file->ReadUInt32(), HashAlt = file->ReadUInt16(), Compression = file->ReadUInt16(), DataLBA = file->ReadUInt32(), DataLength = file->ReadUInt32()};
		return entry;
	}

	void IDXFile::WriteEntry(IDXEntry *entry, unsigned int count = 0)
	{
//C# TO C++ CONVERTER NOTE: The following 'using' block is replaced by its C++ equivalent:
//C# TO C++ CONVERTER TODO TASK: C# to C++ Converter does not resolve named parameters in method calls:
//		using (var bw = new BinaryStream(file.BaseStream, leaveOpen: true))
		BinaryStream *bw = new BinaryStream(file->BaseStream, leaveOpen: true);
		try
		{
			bw->Write(entry->Hash);
			bw->Write(entry->HashAlt);
			bw->Write(entry->Compression);
			bw->Write(entry->DataLBA);
			bw->Write(entry->DataLength);
			if (count != 0)
			{
				bw->getBaseStream()->Position = 0;
				bw->Write(count);
			}
		}
//C# TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to the exception 'finally' clause:
		finally
		{
			if (bw != 0)
				bw.Dispose();
		}
	}

	void IDXFile::FindEntryByHash(unsigned int hash)
	{
		file->BaseStream->Position = 4;
		for (unsigned int i = 0; i < getCount(); ++i)
		{
			if (file->ReadUInt32() == hash)
			{
				file->BaseStream->Position -= 4;
				return;
			}
			file->BaseStream->Position += 12;
		}
		throw new FileNotFoundException();
	}

	void IDXFile::RelinkEntry(unsigned int hash, unsigned int target)
	{
		FindEntryByHash(target);
		file->BaseStream->Position += 4;
		IDXEntry *entry = new IDXEntry {Hash = hash, HashAlt = file->ReadUInt16(), Compression = file->ReadUInt16(), DataLBA = file->ReadUInt32(), DataLength = file->ReadUInt32()};
		FindEntryByHash(hash);
		WriteEntry(entry);
	}

	void IDXFile::ModifyEntry(IDXEntry *entry)
	{
		FindEntryByHash(entry->Hash);
		WriteEntry(entry);
	}

	void IDXFile::AddEntry(IDXEntry *entry)
	{
		file->BaseStream->Position = 4 + 16*getCount();
		setCount(getCount() + 1);
		WriteEntry(entry, getCount());
	}

	void IDXFileWriter::AddEntry(IDXFile::IDXEntry *entry)
	{
		entries.push_back(entry);
	}

	void IDXFileWriter::RelinkEntry(unsigned int hash, unsigned int target)
	{
//C# TO C++ CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted to native C++ unless the option to convert to C++0x lambdas is selected:
		IDXFile::IDXEntry *t = entries.Find(e => e->Hash == target);
		if (t->Hash == 0)
		{
			throw new FileNotFoundException();
		}
		entries.push_back(new IDXFile::IDXEntry {Hash = hash, HashAlt = 0, Compression = t->Compression, DataLBA = t->DataLBA, DataLength = t->DataLength});
	}

	MemoryStream *IDXFileWriter::GetStream()
	{
//C# TO C++ CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted to native C++ unless the option to convert to C++0x lambdas is selected:
		entries.Sort((a, b) => a::Hash < b::Hash ? - 1 : (a::Hash > b::Hash ? 1 : 0));
		MemoryStream *ms = new MemoryStream();
		try
		{
//C# TO C++ CONVERTER NOTE: The following 'using' block is replaced by its C++ equivalent:
//C# TO C++ CONVERTER TODO TASK: C# to C++ Converter does not resolve named parameters in method calls:
//			using (var bw = new BinaryStream(ms, leaveOpen: true))
			BinaryStream *bw = new BinaryStream(ms, leaveOpen: true);
			try
			{
				bw->Write(entries.size());
				for (std::vector<IDXFile::IDXEntry*>::const_iterator entry = entries.begin(); entry != entries.end(); ++entry)
				{
					bw->Write((*entry)->Hash);
					bw->Write((*entry)->HashAlt);
					bw->Write((*entry)->Compression);
					bw->Write((*entry)->DataLBA);
					bw->Write((*entry)->DataLength);
				}
			}
//C# TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to the exception 'finally' clause:
			finally
			{
				if (bw != 0)
					bw.Dispose();
			}
		}
		catch (std::exception &e1)
		{
			ms->Close();
			throw;
		}
		ms->Position = 0;
		return ms;
	}

	IMGFile::IMGFile(Stream *file, long long offset = 0, bool leaveOpen = false) : leaveOpen(leaveOpen), offset(offset)
	{
		this->file = file;
	}

	IMGFile::~IMGFile()
	{
		this->Dispose(true);
	}

	void IMGFile::Dispose(bool disposing)
	{
		if (file != 0)
		{
			if (!leaveOpen)
			{
				delete file;
			}
//C# TO C++ CONVERTER WARNING: C# to C++ Converter converted the original 'null' assignment to a call to 'delete', but you should review memory allocation of all pointer variables in the converted code:
			delete file;
		}
	}

	void IMGFile::ReadFileBuffer(Stream *destination, long long origin, unsigned int length)
	{
		//Not thread safe, but I'm not using threads
		file->Position = offset + origin;
		int num;
		while (length > 0 && (num = file->Read(copyBuffer, 0, static_cast<int>(__min(2048*2, length)))) != 0)
		{
			destination->Write(copyBuffer, 0, num);
			length -= static_cast<unsigned int>(num);
		}
	}

	void IMGFile::EnsureBoundary()
	{
		if (((file->Position - offset) % 2048) != 0)
		{
			int rem = 2048 - static_cast<int>((file->Position - offset) % 2048);
//ORIGINAL LINE: byte[] buf = {0, 0, 0, 0};
//C# TO C++ CONVERTER WARNING: Since the array size is not known in this declaration, C# to C++ Converter has converted this array to a pointer.  You will need to call 'delete[]' where appropriate:
			unsigned char *buf = {0, 0, 0, 0};
			while (rem > 3)
			{
				file->Write(buf, 0, 4);
				rem -= 4;
			}
			while (--rem >= 0)
			{
				file->WriteByte(0);
			}
		}
		if (((file->Position - offset) % 2048) != 0)
		{
			throw new DataMisalignedException();
		}
	}

	GovanifY::Utility::Substream *IMGFile::GetFileStream(IDXFile::IDXEntry *entry)
	{
		return new Substream(file, offset + entry->DataLBA*2048, entry->getIsCompressed() ? entry->getCompressedDataLength() : entry->DataLength);
	}

	void IMGFile::Seek(unsigned int sector)
	{
		file->Position = offset + sector*2048;
	}

	void IMGFile::ReadFile(IDXFile::IDXEntry *entry, Stream *target, bool AdSize)
	{
		if (entry->getIsCompressed())
		{
			if (entry->getCompressedDataLength() > int::MaxValue)
			{
				throw new NotSupportedException("File to big to decompress");
			}
			unsigned char input[entry->CompressedDataLength];
			Seek(entry->DataLBA);
			file->Read(input, 0, static_cast<int>(entry->getCompressedDataLength()));
			try
			{
//ORIGINAL LINE: byte[] output = KH2Compressor.decompress(input, entry.DataLength);
//C# TO C++ CONVERTER WARNING: Since the array size is not known in this declaration, C# to C++ Converter has converted this array to a pointer.  You will need to call 'delete[]' where appropriate:
				unsigned char *output = KH2Compressor::decompress(input, entry->DataLength);
				target->Write(output, 0, sizeof(output) / sizeof(output[0]));
				if (AdSize)
				{
					std::cout << "Size (unpacked): " << sizeof(output) / sizeof(output[0]) << std::endl;
				}
			}
			catch (std::exception &e)
			{
				Program::WriteError(" ERROR: Failed to decompress: " + e.what());
			}
		}
		else
		{
			ReadFileBuffer(target, entry->getOffset(), entry->DataLength);
		}
	}

	void IMGFile::WriteFile(Stream *data)
	{
		if (data->Length > 0xFFFFFFFF)
		{
			throw new NotSupportedException("data too big to store");
		}
		EnsureBoundary();
		data->CopyTo(file);
		EnsureBoundary();
	}

	void IMGFile::AppendFile(IDXFile::IDXEntry *entry, Stream *data)
	{
		if (data->Length > 0xFFFFFFFF)
		{
			throw new NotSupportedException("data too big to store");
		}
		file->Seek(0, SeekOrigin::End);
		EnsureBoundary();
		entry->DataLBA = static_cast<unsigned int>(file->Position - offset) / 2048;
		data->CopyTo(file);
		EnsureBoundary();
	}

	IDXIMGWriter::IDXIMGWriter(Stream *img, long long imgoffset = 0, bool leaveOpen = false)
	{
		InitializeInstanceFields();
		this->setimg(new IMGFile(img, imgoffset));
		this->setimg(new IMGFile(img, imgoffset, leaveOpen));
	}

	const IMGFile &IDXIMGWriter::getimg() const
	{
		return privateimg;
	}

	void IDXIMGWriter::setimg(const IMGFile &value)
	{
		privateimg = value;
	}

	IDXIMGWriter::~IDXIMGWriter()
	{
		InitializeInstanceFields();
		this->Dispose(true);
	}

	void IDXIMGWriter::Dispose(bool disposing)
	{
		if (getimg() != 0)
		{
			delete getimg();
//C# TO C++ CONVERTER WARNING: C# to C++ Converter converted the original 'null' assignment to a call to 'delete', but you should review memory allocation of all pointer variables in the converted code:
			delete getimg();
		}
		if (disposing && idx != 0)
		{
//C# TO C++ CONVERTER WARNING: C# to C++ Converter converted the original 'null' assignment to a call to 'delete', but you should review memory allocation of all pointer variables in the converted code:
			delete idx;
		}
	}

	void IDXIMGWriter::AddFile(IDXFile::IDXEntry *file, Stream *data)
	{
		getimg()->AppendFile(file, data);
		idx->AddEntry(file);
	}

	void IDXIMGWriter::RelinkFile(unsigned int hash, unsigned int target)
	{
		idx->RelinkEntry(hash, target);
	}

	MemoryStream *IDXIMGWriter::GetCurrentIDX()
	{
		return idx->GetStream();
	}

	void IDXIMGWriter::InitializeInstanceFields()
	{
		idx = new IDXFileWriter();
	}
}
